### PR TITLE
[Spark][Sharing] Add util functions for "delta format sharing"

### PR DESCRIPTION
**Pull request by linzhou-db**
#### Which Delta project/connector is this regarding?
- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [x] Other (delta sharing)

## Description
First PR of https://github.com/delta-io/delta/issues/2291: Add util functions for "delta format sharing".
- DeltaSharingUtils: Util functions such as prepare refresh function, etc.
- DeltaSharingFileIndex: Used to support batch queries, to do filtering pushdown
- DeltaSharingLogFileSystem: Used to manage local constructed delta log, handle the read/write with BlockManager.

## How was this patch tested?
Unit Tests

## Does this PR introduce _any_ user-facing changes?
No

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -213,6 +213,31 @@ lazy val contribs = (project in file("contribs"))
     Compile / compile := ((Compile / compile) dependsOn createTargetClassesDir).value
   ).configureUnidoc()
 
+lazy val deltaSharing = (project in file("deltasharing"))
```
Comment (tdas) [inline build.sbt:216]: lets just name the directory sharing. the whole repo is already delta... so does not make sense to have a subdir also named "delta*"

**Comment by tdas** [comment]
Can you a line in the description explaining that this is first of a few items in the issue #2291?

File: deltasharing/src/test/scala/io/delta/sharing/spark/DeltaSharingUtilsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+package io.delta.sharing.spark
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.{SharedSparkContext, SparkEnv, SparkFunSuite}
+import org.apache.spark.storage.BlockId
+
+class DeltaSharingUtilsSuite extends SparkFunSuite with SharedSparkContext {
```
Comment (tdas) [inline deltasharing/src/test/scala/io/delta/sharing/spark/DeltaSharingUtilsSuite.scala:24]: are you sure these tests are running??

Comment (tdas) [reply deltasharing/src/test/scala/io/delta/sharing/spark/DeltaSharingUtilsSuite.scala:24]: @vkorukanti can you validate that these test are running? 
my guess is that they are not, as the build does not add this module to the sparkGroup, and the sparkGroup determines which modules tests are being run under the spark test runner.

there is a choice to be made. 
1. either we add these sharing tests to the sparkGroup. same runner runs these tests but that runner will take longer (its already bad, many hours)
2. or we create a new group and a new test runner.

File: build.sbt [hunk 3]
```diff
@@ -213,6 +213,31 @@ lazy val contribs = (project in file("contribs"))
...
+  .dependsOn(spark % "compile->compile;test->test;provided->provided")
+  .settings(
+    name := "delta-sharing-spark",
+    commonSettings,
+    scalaStyleSettings,
+    javaOnlyReleaseSettings,
```
Comment (tdas) [inline build.sbt:222]: @vkorukanti @linzhou-db can you validate that the maven artifact name is as expected? no scala version in it?

File: build.sbt (hunk 1 above)
Comment (linzhou-db) [reply build.sbt]: updated.

File: build.sbt (hunk 3 above)
Comment (linzhou-db) [reply build.sbt]: no it's not intended, removed `javaOnlyReleaseSettings`.

File: deltasharing/src/test/scala/io/delta/sharing/spark/DeltaSharingUtilsSuite.scala (hunk 2 above)
Comment (linzhou-db) [reply deltasharing/src/test/scala/io/delta/sharing/spark/DeltaSharingUtilsSuite.scala:24]: Good catch. I only manually ran tests. 

let me know the preferred way.

**Review by vkorukanti — APPROVED** [review]
Its a lot of code, but overall lgtm as the pattern followed to read the DeltaLogs is somewhat similar to the reading Parquet files in delta-sharing-client. However, it feels like some of the classes mentioned (response objects of RPCs) be part of the delta-sharing-client lib and not here. This can be improved later in subsequent releases of delta-sharing-client.

**Comment by vkorukanti** [comment]
Verified maven artifacts are generated correctly and tests are run.

File: sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingFileIndex.scala [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+import org.apache.spark.sql.execution.datasources.{FileIndex, PartitionDirectory}
+import org.apache.spark.sql.types.StructType
+
+private[sharing] case class DeltaSharingFileIndexParams(
+    path: Path,
+    spark: SparkSession,
+    metadata: model.DeltaSharingMetadata,
+    options: DeltaSharingOptions)
+
+case class DeltaSharingFileIndex(
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingFileIndex.scala:42]: added

File: build.sbt [hunk 5]
```diff
@@ -213,6 +213,30 @@ lazy val contribs = (project in file("contribs"))
...
+  .dependsOn(spark % "compile->compile;test->test;provided->provided")
+  .settings(
+    name := "delta-sharing-spark",
+    commonSettings,
+    scalaStyleSettings,
+    Test / javaOptions ++= Seq("-ea"),
+    libraryDependencies ++= Seq(
+      "org.apache.spark" %% "spark-sql" % sparkVersion % "provided",
+
+      "io.delta" %% "delta-sharing-client" % "1.0.3",
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply build.sbt:226]: some spark libraries and a org.apache.httpcomponents. 
I will try exclude some of the spark libraries when trying to build the target.

File: sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingLogFileSystem.scala [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+import org.apache.spark.SparkEnv
+import org.apache.spark.internal.Logging
+import org.apache.spark.storage.{BlockId}
+
+/** Read-only file system for delta sharing log.
+ * This is a faked file system to serve data under path delta-sharing-log:/. The delta log will be
+ * prepared by DeltaSharingDataSource and its related classes, put in blockManager, and then serve
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingLogFileSystem.scala:49]: I'll add it in a separate PR, still WIP.

File: sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingUtils.scala [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+import org.apache.spark.SparkEnv
+import org.apache.spark.delta.sharing.TableRefreshResult
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.datasources.FileFormat
+import org.apache.spark.storage.{BlockId, StorageLevel}
+
+object DeltaSharingUtils extends Logging {
+
+  val DELTA_SHARING_BLOCK_ID_PREFIX = "test_delta-sharing"
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingUtils.scala:47]: added

File: sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingUtils.scala [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+  // Refresher function for CachedTableManager to use.
+  // It takes refreshToken: Option[String] as a parameter and return TableRefreshResult.
+  type RefresherFunction = Option[String] => TableRefreshResult
+
+  case class DeltaSharingTableMetadata(
+      version: Long,
+      protocol: model.DeltaSharingProtocol,
+      metadata: model.DeltaSharingMetadata
+  )
+
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingUtils.scala:58]: updated.

File: sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingUtils.scala [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+  private def removeBlockForJsonLogIfExists(blockId: BlockId): Unit = {
+    val blockManager = SparkEnv.get.blockManager
+    blockManager.getMatchingBlockIds(_.name == blockId.name).foreach { b =>
+      logWarning(s"Found and removing existing block for $blockId.")
+      blockManager.removeBlock(b)
+    }
+  }
+
+  val FAKE_CHECKPOINT_FILE_BASE64_ENCODED_STRING =
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply sharing/src/main/scala/io/delta/sharing/spark/DeltaSharingUtils.scala:171]: added comment.

**Pull request closed without merging**